app/controllers/concerns/authenticatable.rb

```diff
@@ -0,0 +1,51 @@
+module Authenticatable
+  SSO_JWT_SESSION_KEY = :sso_session
+  SSO_JWT_REFRESH_KEY = :sso_refresh
+  SSO_JWT_USER_ID_KEY = :sso_user_id
+
+  def skip_authentication
```

Comment:
📝

This method is similar in principle to [require_no_authentication](https://github.com/heartcombo/devise/blob/a259ff3c28912a27329727f4a3c2623d3f5cb6f2/app/controllers/devise_controller.rb#L116).

It short-circuits auth/session related pages and redirects users to the primary logged in page.

`require_no_authentication` is insufficient when you have the following combinations:

logged in via email, visit sso session pages
logged in via sso, visit email session pages
logged in via email, visit email session pages
logged in via sso, visit sso session pages

Basically, if you're logged in as any other login method, and if you visit any kind of login-auth pages, short-circuit.